scipy/stats/_binomtest.py

```diff
@@ -313,16 +312,59 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
             # special case as shortcut, would also be handled by `else` below
             pval = 1.
         elif k < p * n:
-            i = np.arange(np.ceil(p * n), n+1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = n-_binary_search_for_binom_tst(lambda x1: -binom.pmf(x1, n, p),
+                                               -d*rerr, np.ceil(p * n), n)
             pval = binom.cdf(k, n, p) + binom.sf(n - y, n, p)
         else:
-            i = np.arange(np.floor(p*n) + 1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = _binary_search_for_binom_tst(lambda x1: binom.pmf(x1, n, p),
+                                             d*rerr, 0, np.floor(p * n) + 1,
+                                             True)+1
             pval = binom.cdf(y-1, n, p) + binom.sf(k-1, n, p)
 
         pval = min(1.0, pval)
 
     result = BinomTestResult(k=k, n=n, alternative=alternative,
                              proportion_estimate=k/n, pvalue=pval)
     return result
+
+
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
+    r"""
+    Conducts an implicit binary search on a funcion specified by `a.
+
+    Meant to be used on the binomial PMF for the case of two-sided tests
+    to obtain the value on the other side of the mode where the tail prbability
+    should be computed. Because the values past the mode are always in
+    ascending order, this makes binary search applicable.
+
+    Parameters
+    ----------
+    a : callable
+      The function over which to perform binary search. Its values
+      for inputs lo and hi should be in ascending order.
+    d : float
+      The value to search.
+    lo : int
+      The lower end of range to search.
+    hi : int
+      The higher end of the range to search.
+
+    Returns
+    ----------
+    int
+      The index, i between lo and hi
+      such that a(i)<=d<a(i+1)
+    """
+    while lo < hi:
+        mid = lo + (hi-lo)//2
+        midval = a(mid)
+        if midval < d:
+            lo = mid+1
+        elif midval > d:
+            hi = mid-1
+        else:
+            return mid
+    if a(lo) <= d:
+        return lo
+    else:
+        return lo-1
```

Comment:
I'm not clear on why these are necessary, in the tests and code usage there are cases where a `+ 1` is added to the result of the binary search call. Would dropping this block eliminate the need for the `+ 1`s or are these needed for another reason?

Reply:
We need a(x)<d<a(x+1) on termination (x is returned value). I don't think there is a guarantee that when the while terminates, lo will necessarily satisfy this (lo could surpass the d in the loop). We could have a(lo)<d<a(lo+1) or a(lo-1)<d<a(lo). No?

Reply:
I just confirmed both conditions can be exercised depending on inputs. For example, when (9,21,0.48) is passed, the if condition is exercised. When (4,21,0.48) is passed on the other hand, the else condition is exercised.